### ISSUE TITLE
igbinary_unserialize_header: unsupported version

### ISSUE THREAD

**Issue by lngphp** [igbinary/igbinary#20]
Hi!

I have installed igbinary 1.1.1 from PECL, configured it as serializer  for session.serialize_handler    in php.ini so that it is shown in phpinfo(). The change was tested - so far everything is fine, but on live server, I can see warnings like this:

[27-Dec-2012 09:10:23 UTC] PHP Warning:  igbinary_unserialize_header: unsupported version: 1346458177, should be 1 or 2 in /opt/webapps/v1.0.3/includes/dispatcher.php on line 73
[27-Dec-2012 09:10:23 UTC] PHP Warning:  session_start(): Trying to destroy uninitialized session in /opt/webapps/v1.0.3/includes/dispatcher.php on line 73
[27-Dec-2012 09:10:23 UTC] PHP Warning:  session_start(): Failed to decode session object. Session has been destroyed in /opt/webapps/v1.0.3/includes/dispatcher.php on line 73

These messages appear not all the time - session_start() from line 73 worked for me during tests.

What would be the solution? 

PS I have Nginx 1.2.3, php5-fpm 5.3.10-1ubuntu3.4

**Comment by EugenMayer** [comment]
this could be related to #23 (most probably)

**Comment by CybotTM** [comment]
i have the same problem, i get this error for old data serialized with php.

it would be nice if igbinary could fall back to PHP serializer (or any other configurable serializer) if version could not detected.

**Comment by phadej** [comment]
The fallback can be made in the *userland*, in PHP side. We'll improve the warning message to be informative. See #51

**Comment by TysonAndre** [comment]
I haven't actually tested this, but a **userland** method to migrate sessions to PHP would be (See https://github.com/igbinary/igbinary/issues/20#issuecomment-135451416)


(Assuming you have a well tested other interface called SomeOtherSessionHandlerImplementation)

This would just involve converting to igbinary (if necessary) before writing to memcached/redis/files (when session.serialize_handler=igbinary, and converting to igbinary(if necessary) after reading bytes from memcached/redis/files.

```
<?php

class IgbinaryMigratoryHandler implements SessionHandlerInterface {
    /** @var SessionHandlerInterface */
    private $_original;
    public function __construct(SessionHandlerInterface $original) {
        $this->_original = $original;
    }

    public function open($savePath, $sessionName) {
        return $this->_original->open($savePath, $sessionName);
    }

    public function close() {
        return $this->_original->close();
    }
    
    public static function is_igbinary($bytes) {
        return strncmp($bytes, "\x00\x00\x00\x02", 4) === 0;
    }

    public static function coerce_to_expected_serialization($bytes) {
        $expectIgbinary = ini_get('session.serialize_handler') === 'igbinary';
        if (self::is_igbinary($bytes) === $expectIgbinary) { 
            return $bytes;
        }
        if ($expectIgbinary) {
            return igbinary_serialize(unserialize($bytes));
        } else { 
            return serialize(igbinary_unserialize($bytes));
        } 
    }

    public function read($id) {
        return self::coerce_to_expected_serialization($this->_original->read($id));
    }

    public function write($id, $bytes) {
        return $this->_original->write($id, self::coerce_to_expected_serialization($bytes));
    }

    public function destroy($id) {
        return $this->_original->destroy($id);
    }

    public function gc($maxlifetime) {
        return $this->_original->gc($maxlifetime);
    }
}

// ini_set('session.serialize_handler', 'igbinary'); // if needed
session_set_save_handler(new IgbinaryMigratoryHandler(new SomeOtherSessionHandlerImplementation()));
```

References:  https://secure.php.net/manual/en/class.sessionhandlerinterface.php https://secure.php.net/manual/en/function.session-set-save-handler.php

**Comment by TysonAndre** [comment]
the warning message has been changed to be more informative in the master branch, and will be released in 2.0.0

**Comment by TysonAndre** [comment]
2.0.0 was released.

Aside: one session saving implementation (php-memcached) does serialize the serializer type along with the data, but that's an exception (And assumes the list of serializers is known). Redis(phpredis) doesn't.
